### PR TITLE
Migrate to pixi

### PR DESCRIPTION
**Pull request by kklein**
I think that we should add a 'development' section in the docs as a follow-up to this PR; explaining how one can locally
- install the environment
- install and run pre-commit hooks and
- run the tests

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -18,32 +18,31 @@ jobs:
...
-          ref: ${{ github.head_ref }}
-          # needed for 'pre-commit-mirrors-insert-license'
-          fetch-depth: 0
-      - name: Run pre-commit-conda
-        uses: quantco/pre-commit-conda@v1
+          environments: default lint
+      - name: pre-commit
+        run: pixi run pre-commit-run --color=always --show-diff-on-failure
 
   mypy-type-checks:
```
Comment (pavelzw) [inline .github/workflows/ci.yml:28]: When properly checking it in pre-commit (see other comment), we don't need this here as it's already being checked by the pre-commit hook.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,45 +1,81 @@
...
+      # ensure pixi environments are up to date
+      # workaround for https://github.com/prefix-dev/pixi/issues/1482
+      - id: pixi-install
+        name: pixi-install
+        entry: pixi install -e default -e lint
+        language: system
+        always_run: true
+        require_serial: true
+        pass_filenames: false
       - id: insert-license
```
Comment (pavelzw) [inline .pre-commit-config.yaml:14]: i don't think we need this in open source repos

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,45 +1,81 @@
...
-    hooks:
-      - id: prettier-conda
+        name: black-conda
+        entry: pixi run -e lint black
+        language: system
+        require_serial: true
+        types: [python]
+      - id: mypy
+        name: mypy
+        entry: pixi run -e lint mypy
```
Comment (pavelzw) [inline .pre-commit-config.yaml:45]:
```suggestion
        entry: pixi run -e default mypy
```

We need to run mypy in the default environment as this is the environment where all our dependencies stubs are stored

File: pixi.toml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+pip = "*"
+python = ">=3.10"
+setuptools-scm = "*"
+setuptools = ">=61"
+ipykernel = "*"
+ipywidgets = "*"
+lime = "*"
+optuna = "*"
+nbconvert = "*"
+mypy = "1.9.0.*"
```
Comment (pavelzw) [inline pixi.toml:22]:
```suggestion
mypy = "*"
```

mypy is already being locked in `pixi.lock`, `*` should be fine

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -53,33 +52,26 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
```
Comment (pavelzw) [inline .github/workflows/ci.yml:54]:
```suggestion
        os: [ubuntu-latest, macos-latest, windows-latest]
```

File: pixi.toml [hunk 6]
```diff
@@ -0,0 +1,81 @@
+[project]
+name = "metalearners"
+channels = ["conda-forge", "nodefaults"]
```
Comment (pavelzw) [inline pixi.toml:3]:
```suggestion
channels = ["conda-forge"]
```

File: pixi.toml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+black = "*"
+ruff = "*"
+prettier = "*"
+insert-license-header = "*"
+taplo = "*"
+typos = "*"
+docformatter = "*"
+pre-commit-hooks = "*"
+types-setuptools = "*"
+mypy = "*"
```
Comment (pavelzw) [inline pixi.toml:48]:
```suggestion
```

see other comments about mypy

Comment (FrancescMartiEscofetQC) [inline pixi.toml:48]: Do we need `mypy` here if we already have it on the dependencies section?

**Review by FrancescMartiEscofetQC — COMMENTED** [review]
Looking good, I think it's important to add in the README how one can build the environment and the basics for development but maybe we can leave it for another PR.

File: pixi.toml (hunk 7 above)
Comment (FrancescMartiEscofetQC) [reply pixi.toml]: Already updated.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (kklein) [reply .pre-commit-config.yaml:14]: My reading of the situation was that we do in order to insert the following header:
```
# # Copyright (c) QuantCo 2024-2024
# # SPDX-License-Identifier: BSD-3-Clause
```
https://github.com/Quantco/metalearners/blob/main/metalearners/metalearner.py#L1-L2

File: .github/workflows/ci.yml (hunk 1 above)
Comment (kklein) [reply .github/workflows/ci.yml]: Removed it!

**Comment by kklein** [comment]
The Windows tests are failing - I created an issue to follow up on it after this PR:
https://github.com/Quantco/metalearners/issues/24

File: .pre-commit-config.yaml (hunk 2 above)
Comment (pavelzw) [reply .pre-commit-config.yaml:14]: the extra `# ` is definitely too much. not sure whether we need the license header in open-source repos

**Review by pavelzw — APPROVED** [review]

File: .github/workflows/benchmarks.yml [hunk 8]
```diff
@@ -13,20 +13,16 @@ jobs:
...
-      - name: Setup conda
-        uses: mamba-org/setup-micromamba@v1
-        with:
-          condarc-file: .github/assets/.condarc
-          environment-file: benchmarks/environment.yml
-      - name: Install package
-        run: python -m pip install -e .
-      - name: Run benchmark.py
-        run: python benchmarks/benchmark.py
+          environments: default lint
```
Comment (FrancescMartiEscofetQC) [inline .github/workflows/benchmarks.yml:21]: Is there a reason why we need the `lint` environment here?

Comment (kklein) [reply .github/workflows/benchmarks.yml]: Great catch!
[8a43f13](https://github.com/Quantco/metalearners/pull/12/commits/8a43f13d397bded9b8dac8a926262bff5823a924)

**Review by FrancescMartiEscofetQC — APPROVED** [review]
LGTM!

File: .pre-commit-config.yaml (hunk 2 above)
Comment (xhochy) [reply .pre-commit-config.yaml:14]: We should have that, too. It is not as important as for internal things, but we should keep it here.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -18,68 +18,41 @@ jobs:
...
+        if: matrix.python-version == '3.12'
+        uses: codecov/codecov-action@v3.1.3
         with:
-          report-title: Unit Tests - ${{ matrix.os == 'ubuntu-latest' && 'Linux' || 'Windows' }} - Python ${{ matrix.python-version }}
-          custom-arguments: --cov=metalearners --cov-report=xml --cov-report term-missing --color=yes
+          file: ./coverage.xml
       - name: Upload coverage reports to Codecov
-        if: matrix.python-version == '3.12'
-        uses: codecov/codecov-action@125fc84a9a348dbcf27191600683ec096ec9021c
+        if: ${{ matrix.env }} == "py312"
```
Comment (pavelzw) [inline .github/workflows/ci.yml:54]:
```suggestion
        if: matrix.env == "py312"
```

This should work as well

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -1,45 +1,81 @@
...
+        require_serial: true
+        pass_filenames: false
       - id: insert-license
+        name: insert-license
+        entry: pixi run -e lint insert-license
         types: [python]
+        language: system
         args:
           - --license-base64
           - IyBDb3B5cmlnaHQgKGMpIFF1YW50Q28gMjAyNC0yMDI0CiMgU1BEWC1MaWNlbnNlLUlkZW50aWZpZXI6IEJTRC0zLUNsYXVzZQo=
```
Comment (pavelzw) [inline .pre-commit-config.yaml:21]: Please replace the base64 with 

```suggestion
          - Q29weXJpZ2h0IChjKSBRdWFudENvIDIwMjQtMjAyNApTUERYLUxpY2Vuc2UtSWRlbnRpZmllcjogQlNELTMtQ2xhdXNl
```

**Review by pavelzw — APPROVED** [review]

**Review by FrancescMartiEscofetQC — APPROVED** [review]

**Pull request merged**